### PR TITLE
 First pass at a wordsearch solver

### LINKED ISSUE #34: Word search solver
Most of the work is done by https://github.com/JHotterbeekx/word-search-solver, but create a wrapper which takes a dictionary file and a text block.

### PR DESCRIPTION
**Pull request by sjenks**
- Requires https://github.com/kfarnung/DefinitelyTyped/commit/trie-prefix-tree is merged
- Fixes #34 (mostly)

Why not use https://github.com/JHotterbeekx/word-search-solver ?  It only returns first/last letter positions, making highlighting annoying.  It also loops over each letter, direction, and word.  This is more efficient, especially for large word lists (such as the dictionary suggested in #34)

### CODE REVIEW

File: src/WordSearch/WordSearchPoint.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export class WordSearchPoint {
```
Comment (beckbria) [inline src/WordSearch/WordSearchPoint.ts]: "WordSearchPoint" seems fairly verbose given that this is in the Word Search directory.  I'd be tempted to name this Point - then it'll be a simpler refactor if we ever move this into a Common directory or something similar.

File: src/WordSearch/WordSearchPoint.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export class WordSearchPoint {
+  i: number;
```
Comment (beckbria) [inline src/WordSearch/WordSearchPoint.ts]: x/y?

File: src/WordSearch/WordSearchSolver.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      [1, 1], [-1, -1], [1, -1], [-1, 1]
+    ];
+
+    this._trie = trie([]);
+  }
+
+  findWords(words: string[]): WordSearchResult[] {
+    for (const word of words) {
+      if (word !== '') {
+        this._trie.addWord(word);
```
Comment (beckbria) [inline src/WordSearch/WordSearchSolver.ts]: Consider trimming whitespace.  You could also think about validating that all characters are in an expected set (alphanumeric?), but in the name of i18n that may have edge cases I haven't thought about.

File: test/wordsearch.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+/* global describe, it */
+
+const assert = require('assert');
+const { WordSearchSolver } = require('../');
+
+describe('WordSearchSolver', function () {
+  it('Cardnial directions', function () {
```
Comment (beckbria) [inline test/wordsearch.js]: Nit: Cardinal

File: src/WordSearch/WordSearchResult.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import {Point} from './WordSearchPoint';
+
+export class WordSearchResult {
+  word: string;
+  points: Point[];
+  constructor(word: string, wordPoints: Point[]) {
+    this.word = word;
+    this.points = wordPoints;
```
Comment (beckbria) [inline src/WordSearch/WordSearchResult.ts]: You probably want a deep copy here to prevent the caller from messing with this array after passing it to you.  I believe that right now this just copies a reference to the array, not the elements.  @kfarnung does that seem correct?

File: src/WordSearch/WordSearchResult.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import {Point} from './Point';
+
+export class WordSearchResult {
```
Comment (kfarnung) [inline src/WordSearch/WordSearchResult.ts]: nit: I would just call this Result and rename it on export.

File: src/WordSearch/WordSearchResult.ts (hunk 5 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchResult.ts]: That's not a bad idea, it might be too new, but `Array.from(other)` should make a shallow copy of the elements.  I don't think it would be too surprising to most devs that modifying the array after passing it could blow up, but might as well harden the surface and if the copy ever becomes a problem, deal with it later.

File: src/WordSearch/Point.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export class Point {
+  x: number;
```
Comment (kfarnung) [inline src/WordSearch/Point.ts]: Should these be mutable?  Would it be worth adding `readonly` to at least harden the TypeScript surface?

File: src/WordSearch/WordSearchSolver.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      // Diagonal
+      [1, 1], [-1, -1], [1, -1], [-1, 1]
+    ];
+
+    this._targets = trie([]);
+  }
+
+  findWords(words: string[]): WordSearchResult[] {
+    for (const word of words) {
+      if (word !== '') {
```
Comment (kfarnung) [inline src/WordSearch/WordSearchSolver.ts]: What about null, undefined, numbers, etc?  Likely we should throw in cases where an invalid input is passed, right?

File: src/WordSearch/WordSearchSolver.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        this._targets.addWord(word.trim());
+      }
+    }
+    return this.search();
+  }
+
+  private search(): WordSearchResult[] {
+    let results: WordSearchResult[] = [];
+    for (let y = 0; y < this._matrix.length; y++) {
```
Comment (kfarnung) [inline src/WordSearch/WordSearchSolver.ts]: nit: At least at one point it was better practice to pull the length out to a local before the loop to avoid unnecessary reads on each iteration.

File: src/WordSearch/WordSearchSolver.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      for (let x = 0; x < this._matrix[y].length; x++) {
+        const point = new Point(x, y);
+        const pointResults = this.startSearch(point);
+        results = results.concat(pointResults);
```
Comment (kfarnung) [inline src/WordSearch/WordSearchSolver.ts]: You likely want to push onto the original array rather than creating a new array for each, I think you can use "spread args" to do it, e.g.:

```
results.push(...pointResults);
```

or at worst case:

```
results.push.apply(results, pointResults)
```

File: src/trie-prefix-tree.d.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
+// Type definitions for trie-prefix-tree 1.5
+// Project: https://github.com/lyndseybrowning/trie-prefix#readme
+// Definitions by: James Lismore <https://github.com/jlismore>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
+declare module 'trie-prefix-tree'
```
Comment (kfarnung) [inline src/trie-prefix-tree.d.ts]: Add a blank line and a comment explaining why this file is necessary.

File: test/wordsearch.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    const results = solver.findWords(['puzz', 'baz']);
+
+    assert.strictEqual(results.length, 2);
+  });
+
+  it('Jagged array outside box', function () {
+    const matrix = [
+      [ 'x', 'f', 'o', 'o' ],
+      [ 'b', 'x', 'x', ],
+      [ 'a', 'x', 'b', 'a', 'r' ],
```
Comment (kfarnung) [inline test/wordsearch.js]: Can you have this one go backwards to ensure that we're checking for words starting outside the non-jagged part of the array?

File: test/wordsearch.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    const matrix = [
+      [ 'p', 'x', 'x', 'x' ],
+      [ 'u', 'w', 'i', 'n' ],
+      [ 'z', 'x', 'x', 'x' ],
+      [ 'z', 'x', 'x', 'x' ]
+    ];
+    const solver = new WordSearchSolver(matrix);
+    const results = solver.findWords(['puzz', 'win', 'foo', 'bar', 'baz']);
+
+    assert.strictEqual(results.length, 2);
```
Comment (kfarnung) [inline test/wordsearch.js]: I wonder if it would be worth adding a couple more asserts that verify that the results array contains the exact words we're expecting.

File: test/wordsearch.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+      [ 'x', 'x', 'x', 'x' ],
+      [ 'x', 'x', 'x', 'x' ],
+    ];
+    const solver = new WordSearchSolver(matrix);
+    const results = solver.findWords(['foo', 'bar', 'baz']);
+
+    assert.strictEqual(results.length, 1);
+    assert.strictEqual(results[0].word, 'foo');
+
+    assert.strictEqual(results[0].points[0].x, 1);
```
Comment (kfarnung) [inline test/wordsearch.js]: You should also assert the length of the points array for good measure.

Comment (kfarnung) [reply test/wordsearch.js]: You could also use array spread to potentially make this cleaner:

```
const [ point0, point1, point2 ] = results[0].points;
```

File: src/WordSearch/WordSearchSolver.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+import {Point} from './Point';
+import {WordSearchResult} from './WordSearchResult';
+
+export class WordSearchSolver {
+  // The grid of letters which makes up the wordsearch puzzle
+  private _matrix: string[][];
+  // Transpositions to apply when finding words
+  private _directions: number[][];
+  // Potential words to find in the letter grid
+  private _targets: any;  // tslint:disable-line
```
Comment (kfarnung) [inline src/WordSearch/WordSearchSolver.ts]: I tend to prefer:

```
// tslint:disable-next-line:no-any
```

Rather than the same line, turn everything off approach.

File: test/wordsearch.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
+/* global describe, it */
+
+const assert = require('assert');
+const { WordSearchSolver } = require('../');
+
+describe('WordSearchSolver', function () {
```
Comment (kfarnung) [inline test/wordsearch.js]: I don't see a test for the case where there are two words in the list which overlap in the same direction, e.g. "foot" and "football".  It looks like the code handles it, but we should probably validate it.

File: test/wordsearch.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    const solver = new WordSearchSolver(matrix);
+    const results = solver.findWords(['puzz', 'baz']);
+
+    assert.strictEqual(results.length, 2);
+  });
+
+  it('Jagged array outside box', function () {
+    const matrix = [
+      [ 'x', 'f', 'o', 'o' ],
+      [ 'b', 'x', 'x', ],
```
Comment (kfarnung) [inline test/wordsearch.js]: The semistandard linter doesn't seem to like comma dangles and there are three of them in this file.  I pushed a PR to re-enable that linter.

These should be easily fixed with `npx semistandard --fix` from the root of the repo.

File: test/wordsearch.js (hunk 12 above)
Comment (sjenks) [reply test/wordsearch.js]: I think the next test covers this case?

File: src/trie-prefix-tree.d.ts (hunk 11 above)
Comment (sjenks) [reply src/trie-prefix-tree.d.ts]: I think your cherry-picked commit includes this

File: src/WordSearch/WordSearchSolver.ts (hunk 8 above)
Comment (sjenks) [reply src/WordSearch/WordSearchSolver.ts]: I'll add a check for null, but I'd like to keep this as open as possible.  I think as soon as I restrict it to alphabetic characters, the next puzzle is going to be wordsearch of just numbers :P

File: src/WordSearch/WordSearchSolver.ts [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+      // Diagonal
+      [1, 1], [-1, -1], [1, -1], [-1, 1]
+    ];
+
+    this._targets = trie([]);
+  }
+
+  findWords(words: string[]): Result[] {
+    for (const word of words) {
+      if (word === null || typeof (word) === 'undefined') {
```
Comment (kfarnung) [inline src/WordSearch/WordSearchSolver.ts]: You shouldn't need to do typeof, `word === undefined` should work fine

File: src/WordSearch/WordSearchSolver.ts (hunk 8 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: Do we really want to silently ignore empty strings? What about a string containing only spaces? I would argue that you want to trim first and throw an error in the empty case.

File: src/WordSearch/WordSearchSolver.ts (hunk 10 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: I don't see anything resolved here?

File: src/WordSearch/WordSearchSolver.ts (hunk 8 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: I also was referring to the literal type number, it seems like this code assumes a string so if something isn't a string I don't know what the expectation is.

File: test/wordsearch.js (hunk 13 above)
Comment (kfarnung) [reply test/wordsearch.js]: Generally I'd recommend asserting the length before going further, also you should be able to simply do `results.indexOf('puzz') >= 0`.  It's likely faster than iterating yourself.

File: src/WordSearch/WordSearchSolver.ts (hunk 18 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: Also a nit, `typeof` is an operator, not a function, so that's why the parens are weird, typically you wouldn't bother with them at all.

File: src/WordSearch/WordSearchSolver.ts (hunk 10 above)
Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: Ideally you want to have `const results: Results[] = []` and not use `concat` at all.

Comment (kfarnung) [reply src/WordSearch/WordSearchSolver.ts]: Looks like you resolved it below, just want to do the same thing here.

**Comment by kfarnung** [comment]
LGTM with some really minor fixups, squash merge when ready.

File: test/wordsearch.js [hunk 19]
```diff
@@ -0,0 +1,112 @@
+/* global describe, it */
+
+const assert = require('assert');
+const { WordSearchSolver } = require('../');
+
+function assertResultsContainsWord (results, word) {
+  for (const result of results) {
```
Comment (kfarnung) [inline test/wordsearch.js]: `result.indexOf(word) >= 0` should do the same thing, but it's a nit at best.

File: src/WordSearch/WordSearchSolver.ts (hunk 10 above)
Comment (sjenks) [reply src/WordSearch/WordSearchSolver.ts]: Oh, yeah, sorry.  I didn't see that I had done this more than once

File: test/wordsearch.js (hunk 19 above)
Comment (kfarnung) [reply test/wordsearch.js]: Whoops, fair enough.
